File: trezor_protector/trezor.py
"""
Trezor device interface — compatible with trezorlib 0.20+.

Key changes in 0.20.x vs 0.13.x:
  - TrezorClient is now abstract; concrete clients come from get_client()
  - PIN / button callbacks are passed to AppManifest, not a UI object
  - CipherKeyValue is misc.encrypt_keyvalue / decrypt_keyvalue, both take a Session
  - A Session is obtained from get_default_session(client) after the device is unlocked

Flow
----
1. connect(pin_callback, button_callback)
     → AppManifest + get_client() → TrezorClientV1 or TrezorClientThp
2. encrypt_master_key(raw_key)
     → get_default_session(client)  ← device PIN is entered here if needed
     → misc.encrypt_keyvalue(session, …)  ← device button confirmation here
3. decrypt_master_key(enc_key)
     → same pattern, decrypt_keyvalue
"""

from __future__ import annotations

from typing import Callable, Optional

_DERIV_PATH = "m/10016'/0'"
_APP_KEY    = "TrezorProtector"


class TrezorManager:
    """Wraps the trezorlib 0.20+ client with the operations we need."""

    def __init__(self) -> None:
        self._client = None

    # ------------------------------------------------------------------
    # Connection
    # ------------------------------------------------------------------

    def connect(
        self,
        pin_callback:    Optional[Callable] = None,
        button_callback: Optional[Callable] = None,
    ) -> None:
        """
        Open the first available Trezor device.

        For CLI use, pass nothing — ClickUI provides terminal prompts.
        For GUI use, pass callables:
          pin_callback(req: messages.PinMatrixRequest) -> str
          button_callback(req: messages.ButtonRequest)  -> None
        """
        try:
            from trezorlib.client import AppManifest, get_client
            from trezorlib.transport import get_transport

            if pin_callback is None:
                from trezorlib.cli.ui import ClickUI
                _ui = ClickUI()
                pin_callback    = _ui.get_pin
                button_callback = _ui.button_request

            app = AppManifest(
                app_name="TrezorProtector",
                pin_callback=pin_callback,
                button_callback=button_callback,
            )
            transport = get_transport()
            self._client = get_client(app, transport)
        except Exception as exc:
            self._client = None
            raise ConnectionError(
                f"Cannot connect to Trezor: {exc}\n"
                "Make sure the device is plugged in and the Trezor Bridge\n"
                "(or HID/WebUSB drivers on Windows) is installed."
            ) from exc

    def disconnect(self) -> None:
        if self._client is not None:
            try:
                self._client.close()
            except Exception:
                pass
            self._client = None

    # ------------------------------------------------------------------
    # Device info
    # ------------------------------------------------------------------

    def get_info(self) -> dict:
        self._require()
        f = self._client.features
        model = getattr(f, "model", None) or "Trezor One"
        label = getattr(f, "label", None) or "(no label)"
        fw    = f"{f.major_version}.{f.minor_version}.{f.patch_version}"
        return {
            "model":       str(model),
            "label":       str(label),
            "firmware":    fw,
            "initialized": bool(f.initialized),
        }

    # ------------------------------------------------------------------
    # Key operations  (CipherKeyValue via trezorlib 0.20+ API)
    # ------------------------------------------------------------------

    def encrypt_master_key(self, raw_key: bytes) -> bytes:
        """
        Encrypt a 32-byte master key on the device.

        The device will display the app name and ask for button confirmation.
        Returns 32 bytes that are safe to persist on disk.
        """
        if len(raw_key) != 32:
            raise ValueError("raw_key must be exactly 32 bytes")
        self._require()
        return self._keyvalue(raw_key, encrypt=True)

    def decrypt_master_key(self, encrypted_key: bytes) -> bytes:
        """
        Decrypt 32 bytes back to the original master key.

        Requires the same Trezor device (same seed) that originally encrypted it.
        """
        if len(encrypted_key) != 32:
            raise ValueError("encrypted_key must be exactly 32 bytes")
        self._require()
        return self._keyvalue(encrypted_key, encrypt=False)

    def _keyvalue(self, value: bytes, *, encrypt: bool) -> bytes:
        from trezorlib import tools
        from trezorlib.client import get_default_session
        from trezorlib.misc import decrypt_keyvalue, encrypt_keyvalue

        address_n = tools.parse_path(_DERIV_PATH)
        # get_default_session calls client.ensure_unlocked() which triggers PIN
        session = get_default_session(self._client)
        fn = encrypt_keyvalue if encrypt else decrypt_keyvalue
        return fn(
            session,
            address_n,
            _APP_KEY,
            value,
            ask_on_encrypt=True,
            ask_on_decrypt=True,
        )

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    def _require(self) -> None:
        if self._client is None:
            raise RuntimeError("Trezor not connected. Call connect() first.")

    def __enter__(self) -> "TrezorManager":
        self.connect()
        return self

    def __exit__(self, *_) -> None:
        self.disconnect()

File: trezor_protector/rust/tp-core/src/trezor.rs
//! Trezor device access via the official `trezor-client` crate.
//!
//! The vault master key is wrapped with the device's CipherKeyValue
//! operation on the SLIP-0016 path m/10016'/0'. The key string and flags
//! are identical to the Python v1 implementation, so a vault created there
//! unwraps with the same physical device here.
//!
//! Every wrap/unwrap requires an explicit button confirmation on the device
//! (`ask_on_encrypt` / `ask_on_decrypt` are both set), so malware that grabs
//! the vault file cannot silently ask the device for the master key.

use trezor_client::client::common::TrezorResponse;
use trezor_client::client::Trezor;
use trezor_client::protos;

use crate::crypto::SecretKey;
use crate::error::{Error, Result};

/// SLIP-0016 style path m/10016'/0'
const DERIVATION_PATH: [u32; 2] = [10016 | 0x8000_0000, 0x8000_0000];
/// Must stay in sync with the Python v1 app key — it is mixed into the
/// CipherKeyValue derivation, so changing it changes the wrapping key.
const APP_KEY: &str = "TrezorProtector";

/// How the caller answers device interaction requests.
///
/// The CLI prompts on the terminal; the native-messaging host relays the
/// request to the extension popup.
pub trait Interaction {
    /// A PIN matrix is displayed on the device; return the *positions* the
    /// user selected (digits 1-9 as shown on the classic scrambled grid).
    fn pin(&mut self) -> Result<String>;
    /// A passphrase was requested. Return `None` to enter it on the device
    /// itself (always preferred — the host machine never sees it).
    fn passphrase(&mut self) -> Result<Option<String>>;
    /// The device is waiting for a physical button press.
    fn notify_button(&mut self);
}

pub struct DeviceInfo {
    pub model: String,
    pub label: String,
    pub firmware: String,
    pub initialized: bool,
}

pub struct TrezorManager {
    client: Trezor,
}

impl TrezorManager {
    /// Connect to the first available Trezor device.
    pub fn connect() -> Result<Self> {
        let devices = trezor_client::find_devices(false);
        let device = devices.into_iter().next().ok_or_else(|| {
            Error::Trezor(
                "no Trezor device found — plug it in and make sure the WinUSB \
                 driver (installed with Trezor Suite) is present"
                    .into(),
            )
        })?;
        let mut client = device
            .connect()
            .map_err(|e| Error::Trezor(format!("cannot open device: {e}")))?;
        client
            .init_device(None)
            .map_err(|e| Error::Trezor(format!("device initialization failed: {e}")))?;
        Ok(Self { client })
    }

    pub fn info(&self) -> Result<DeviceInfo> {
        let f = self
            .client
            .features()
            .ok_or_else(|| Error::Trezor("device features unavailable".into()))?;
        // The firmware reports terse model codes ("1", "T", "T2B1", …).
        let model = match f.model() {
            "" => "Trezor".to_string(),
            "1" => "Trezor One".to_string(),
            "T" => "Trezor Model T".to_string(),
            "T2B1" | "Safe 3" => "Trezor Safe 3".to_string(),
            "T3T1" | "Safe 5" => "Trezor Safe 5".to_string(),
            other => format!("Trezor {other}"),
        };
        Ok(DeviceInfo {
            model,
            label: if f.label().is_empty() { "(no label)".into() } else { f.label().into() },
            firmware: format!(
                "{}.{}.{}",
                f.major_version(),
                f.minor_version(),
                f.patch_version()
            ),
            initialized: f.initialized(),
        })
    }

    /// Wrap a 32-byte master key on the device. The result is safe to store
    /// on disk: it can only be unwrapped by the same seed + passphrase.
    pub fn encrypt_master_key(
        &mut self,
        raw_key: &SecretKey,
        interaction: &mut dyn Interaction,
    ) -> Result<Vec<u8>> {
        self.cipher_key_value(raw_key.as_bytes().to_vec(), true, interaction)
    }

    /// Unwrap a previously wrapped master key.
    pub fn decrypt_master_key(
        &mut self,
        wrapped: &[u8],
        interaction: &mut dyn Interaction,
    ) -> Result<SecretKey> {
        if wrapped.len() != 32 {
            return Err(Error::Trezor("wrapped master key must be 32 bytes".into()));
        }
        let mut value = self.cipher_key_value(wrapped.to_vec(), false, interaction)?;
        let key = SecretKey::from_slice(&value)?;
        // Wipe the intermediate buffer.
        value.iter_mut().for_each(|b| *b = 0);
        Ok(key)
    }

    fn cipher_key_value(
        &mut self,
        value: Vec<u8>,
        encrypt: bool,
        interaction: &mut dyn Interaction,
    ) -> Result<Vec<u8>> {
        let mut req = protos::CipherKeyValue::new();
        req.address_n = DERIVATION_PATH.to_vec();
        req.set_key(APP_KEY.to_owned());
        req.set_value(value);
        req.set_encrypt(encrypt);
        req.set_ask_on_encrypt(true);
        req.set_ask_on_decrypt(true);

        let resp: TrezorResponse<'_, protos::CipheredKeyValue, protos::CipheredKeyValue> = self
            .client
            .call(req, Box::new(|_, m| Ok(m)))
            .map_err(|e| Error::Trezor(format!("CipherKeyValue failed: {e}")))?;

        let result = drive(resp, interaction)?;
        Ok(result.value().to_vec())
    }
}

/// Drive a device interaction to completion, answering PIN / passphrase /
/// button requests through the supplied handler.
fn drive<'a, T, R: trezor_client::TrezorMessage>(
    mut resp: TrezorResponse<'a, T, R>,
    interaction: &mut dyn Interaction,
) -> Result<T> {
    loop {
        resp = match resp {
            TrezorResponse::Ok(value) => return Ok(value),
            TrezorResponse::Failure(f) => {
                return Err(Error::Trezor(format!(
                    "device refused: {}",
                    f.message()
                )));
            }
            TrezorResponse::ButtonRequest(req) => {
                interaction.notify_button();
                req.ack()
                    .map_err(|e| Error::Trezor(format!("button ack failed: {e}")))?
            }
            TrezorResponse::PinMatrixRequest(req) => {
                let pin = interaction.pin()?;
                if !pin.chars().all(|c| ('1'..='9').contains(&c)) || pin.is_empty() {
                    return Err(Error::InvalidInput(
                        "PIN must be 1-9 matrix positions".into(),
                    ));
                }
                req.ack_pin(pin)
                    .map_err(|e| Error::Trezor(format!("PIN rejected: {e}")))?
            }
            TrezorResponse::PassphraseRequest(req) => match interaction.passphrase()? {
                Some(phrase) => req
                    .ack_passphrase(phrase)
                    .map_err(|e| Error::Trezor(format!("passphrase failed: {e}")))?,
                None => req
                    .ack(true)
                    .map_err(|e| Error::Trezor(format!("passphrase failed: {e}")))?,
            },
        };
    }
}
